chisel/src/options.rs

```diff
@@ -0,0 +1,118 @@
+
+/// Key-value structure for immutable CLI options. Used for storing utility options and
+/// configurations in oneliner mode.
+pub struct ChiselFlags(HashMap<String, String>);
+
+impl ChiselFlags {
+    /// Sets the value of `key`.
+    pub fn set(&mut self, key: &str, value: &str) {
+        self.0.insert(key.to_string(), value.to_string());
+    }
+
+    /// Gets the value of `key`.
+    pub fn value_of(&self, key: &str) -> Option<&str> {
+        match self.0.get(key) {
+            Some(s) => Some(s.as_str()),
+            None => None,
+        }
+    }
+
+    /// Compares the value of `key` to the passed `val`.
+    pub fn value_eq(&self, key: &str, val: &str) -> bool {
+        match self.value_of(key) {
+            Some(s) => s == val,
+            None => false,
+        }
+    }
+
+    /// Apply all flags passed from CLI
+    pub fn apply(&mut self, matches: &ArgMatches) {
+        if matches.is_present("NO_RECOVER") {
+            self.set("util.norecover", "true");
+        }
+        if matches.is_present("DEBUG_MESSAGES") {
+            self.set("util.debugging", "true");
+        }
+        if let Some(value) = matches.value_of("CONFIG") {
+            self.set("run.config.path", value);
+        }
+        if let Some(values) = matches.values_of("MODULES") {
+            let values_collected = values.fold(String::new(), |mut acc, val| {
+                acc.push_str(&format!("{},", val));
+                acc
+            });
+            self.set("oneliner.modules", &values_collected);
+        }
+        if let Some(values) = matches.values_of("MODULE_OPTIONS") {
+            let values_collected = values.fold(String::new(), |mut acc, val| {
+                acc.push_str(&format!("{},", val));
+                acc
+            });
+            self.set("oneliner.modules.options", &values_collected);
+        }
+        if let Some(value) = matches.value_of("FILE") {
+            self.set("oneliner.file", value);
+        }
+        if let Some(value) = matches.value_of("OUTPUT_PATH") {
+            self.set("oneliner.out", value);
+        }
+        if let Some(value) = matches.value_of("OUTPUT_MODE") {
+            match value {
+                val @ "wasm" | val @ "wat" | val @ "hex" => {
+                    self.set("output.mode", val);
+                }
+                _ => panic!("CLI parser only accepts 'wasm', 'wat', or 'hex'"),
+            }
+        }
+    }
+}
+
+impl Deref for ChiselFlags {
+    type Target = HashMap<String, String>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for ChiselFlags {
+    fn default() -> Self {
+        let opts: HashMap<String, String> = HashMap::new();
+        let mut ret = ChiselFlags(opts);
+
+        ret.set("util.norecover", "false");
+        ret.set("util.debugging", "false");
+        ret.set("output.mode", "wasm");
+        ret.set("run.config.path", CHISEL_DEFAULT_CONFIG_PATH);
```

Comment:
I know this is const so it doesn't make a difference, but since the other defaults are here it may make sense just having the value here too.